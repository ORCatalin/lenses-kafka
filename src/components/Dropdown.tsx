import React, { useState, useCallback } from 'react';
import getDeepObjectValue from 'lodash.get';
import classNames from "classnames";

export interface IOption {
  label: string;
  value: string;
}

export interface IDropdown {
  title: string;
  options: IOption[];
  onChange: (value: IOption) => void;
  value: IOption;
}

export default function Dropdown({ title, options, onChange, value }: IDropdown) {
  const [active, setActive] = useState(false);
  const className = classNames('dropdown', { 'is-active': active });
  const handleDDLClick = useCallback(() => setActive(active => !active), []);
  const handleChange = useCallback((event) => {
    const target = event.target;
    const label = target.getAttribute('data-label');
    const value = target.getAttribute('data-value');
    onChange({ label, value });
    setActive(false);
  }, [])

  return (
    <div className={className}>
      <div className="dropdown-trigger">
        <button
          className="button"
          aria-haspopup="true"
          aria-controls="dropdown-menu"
          onClick={handleDDLClick}
        >
          <span>{getDeepObjectValue(value, 'label') ? value.label : title}</span>
          <span className="icon is-small">
        <i className="fas fa-angle-down" aria-hidden="true"/>
      </span>
        </button>
      </div>
      <div className="dropdown-menu" id="dropdown-menu" role="menu">
        <div className="dropdown-content" style={{ overflowY: 'auto', maxHeight: 200 }}>
          {options.map((option, idx) => {
            const className = classNames(
              'dropdown-item',
              { 'is-active': option.value === value.value },
            );
            return (
              <div
                data-value={option.value}
                data-label={option.label}
                className={className}
                key={option.value}
                onClick={handleChange}
              >
                {option.label}
              </div>
            )
          })}
        </div>
      </div>
    </div>
  )
}