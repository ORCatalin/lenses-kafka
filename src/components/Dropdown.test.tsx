import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import Dropdown, { IOption } from "./Dropdown";

const options: IOption[] = [{ label: "1", value: "2" }, { label: "3", value: "4" }];

const renderDropdown = () => render(
  <Dropdown title="Test" options={options} value={options[1]} onChange={() => {
  }}/>
)

describe('Dropdown component', () => {
  test('is rendering', () => {
    const { asFragment } = renderDropdown();
    expect(asFragment()).toMatchSnapshot();
  })
  test('has active option', () => {
    const { container } = renderDropdown();
    expect(container.getElementsByClassName('dropdown-item is-active').length).toBe(1)
  })
  test('opens on click', () => {
    const { container } = renderDropdown();
    fireEvent.click(container.getElementsByTagName('button')[0]);
    expect(container.getElementsByClassName('dropdown is-active').length).toBe(1);
  });
});

