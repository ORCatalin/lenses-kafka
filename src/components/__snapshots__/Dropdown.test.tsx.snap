// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component is rendering 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      class="dropdown-trigger"
    >
      <button
        aria-controls="dropdown-menu"
        aria-haspopup="true"
        class="button"
      >
        <span>
          3
        </span>
        <span
          class="icon is-small"
        >
          <i
            aria-hidden="true"
            class="fas fa-angle-down"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      id="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
        style="overflow-y: auto; max-height: 200px;"
      >
        <div
          class="dropdown-item"
          data-label="1"
          data-value="2"
        >
          1
        </div>
        <div
          class="dropdown-item is-active"
          data-label="3"
          data-value="4"
        >
          3
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
